### PR TITLE
[CANN]: Add host buffer type for Ascend NPU

### PR DESCRIPTION
**Pull request by bachelor-dou**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High
  
This PR adds host buffer type for Ascend NPU, Using ping memory will make it more efficiency. #9304

### CODE REVIEW

**Review by slaren — APPROVED** [review]

**Comment by hipudding** [comment]
@Dou-Git Do not add two different commits in one pull request.

**Comment by Dou-Git** [comment]
> @Dou-Git Do not add two different commits in one pull request.

Okay, my mistake. Thanks.

File: ggml/src/ggml-cann.cpp [hunk 1]
```diff
@@ -1220,6 +1220,73 @@ ggml_backend_cann_buffer_type(int32_t device) {
...
+    ACL_CHECK(aclrtFreeHost(buffer->context));
+}
+
+static void * ggml_cann_host_malloc(size_t size) {
+    if (getenv("GGML_CANN_NO_PINNED") != nullptr) {
+        return nullptr;
+    }
+
+    void * ptr = nullptr;
+    aclError err = aclrtMallocHost((void **) &ptr, size);
```
Comment (hipudding) [inline ggml/src/ggml-cann.cpp:1247]: Use ACL_CHECK

File: ggml/src/ggml-cann.cpp [hunk 2]
```diff
@@ -1220,6 +1220,73 @@ ggml_backend_cann_buffer_type(int32_t device) {
     return &ggml_backend_cann_buffer_types[device];
 }
 
+// host buffer type
```
Comment (hipudding) [inline ggml/src/ggml-cann.cpp:1223]: Add comment in Doxygen format, for all new added functions.

Comment (Dou-Git) [reply ggml/src/ggml-cann.cpp:1223]: OK, got it.

**Pull request merged**